## What are the types of apps that attract the most users in Google Play and App Store?

*To do this, we need to collect and analyse data about mobile apps available on Google Play and the App Store.*

*As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.*

*We have two data sets:*
    *1. A data set containing data about approximately __7,000 IOS apps__ from the __App Store__, the data was collected in __July 2017__;* 
    *2. A data set containing data about approximately __10,000 Android apps__ from __Google Play__, the data was collected in __August 2018__.*
    

In [1]:
def explore(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row)   #printeaza fiecare row dintr-un anumit slice
        print('\n')
        
    if rows_and_columns:   # if True, afiseaza nr de coloane si randuri
        print('Number of rows: ', len(dataset))
        print('Number of columns: ', len(dataset[0]))

In [2]:
open_file_Apple = open('AppleStore.csv', encoding="utf8")
open_file_Google = open('GooglePlayStore.csv',encoding="utf8")

from csv import reader
read_file_Apple = reader(open_file_Apple)
read_file_Google = reader(open_file_Google)

Apple_data = list(read_file_Apple)
Google_data = list(read_file_Google)

__The header in the App Store dataset:__

In [3]:
ios_header = Apple_data[0]
print(ios_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [4]:
print('The number of rows in the Apple Store dataset:', len(Apple_data))
print('The number of columns in the Apple Store dataset:', len(Apple_data[0]))

The number of rows in the Apple Store dataset: 7198
The number of columns in the Apple Store dataset: 16


__The header in the Google Store dataset:__

In [5]:
android_header = Google_data[0]
print(android_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [6]:
print('The number of rows in the Google Store dataset', len(Google_data))
print('The number of columns in the Google Store dataset', len(Google_data[0]))

The number of rows in the Google Store dataset 10842
The number of columns in the Google Store dataset 13


It seems so far that we have 7198 IOS apps in the Apple Store, and 10842 Android apps in the Google Store. 

Now, since we are interested only in the apps that are __free__ and directed toward an __English speaking audience__, we will have to make sure that our our datasets reflect that. 

## Cleaning Data
_Removing or correcting wrong data, removing duplicate data and modifying data to fit the purpose of our analysis._

One of the users has noticed an error to one of the rows in the data, that we need to explore. Link, [here](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015). 
It seems that the issue is in the Google Store dataset. 

In [7]:
print(Google_data[10473])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [8]:
print(android_header)
print(Google_data[10473])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [9]:
explore(Google_data,10470, 10475, False)

['TownWiFi | Wi-Fi Everywhere', 'COMMUNICATION', '3.9', '2372', '58M', '500,000+', 'Free', '0', 'Everyone', 'Communication', 'August 2, 2018', '4.2.1', '4.2 and up']


['Jazz Wi-Fi', 'COMMUNICATION', '3.4', '49', '4.0M', '10,000+', 'Free', '0', 'Everyone', 'Communication', 'February 10, 2017', '0.1', '2.3 and up']


['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']




Having checked the data row flagged by the user against the android header and other data slices, we have noticed that the dataset at the 10473th row is indeed not correct as it doesn't have category attached. 

We will delete that row. 

In [14]:
print(len(Google_data))
#del(Google_data[10473]) # This has been run, the wrong data point has been deleted, proof below 
print(len(Google_data))

10842
10841


__*Now let's inspect whether there are any duplicates in our datasets.
We will create a dictionary that maintains all the unique values that we will be comparing against.*__

# 1. Apple Store 
## Identify Duplicates
    

In [11]:
print(len(Apple_data))
print(Apple_data[1][0])
print(Apple_data[1][1])

7198
284882215
Facebook


In [38]:
#print(len(Apple_data)-1) # -1 because we don't count the header in

def frequency_var(dataset, column): #create a function to identify duplicates

    frequency_table = {} 

    for row in dataset[1:]:
        var = row[column]    
        if var in frequency_table:
            frequency_table[var] += 1
        else:
            frequency_table[var] = 1
    
    duplicate_list = []
    
    if len(frequency_table) != (len(dataset)-1): #daca intr-adevar exista duplicate, trecem sa cautam care sunt acestea
        for i in frequency_table:
            if frequency_table[i] != 1:
                duplicate_list.append(i) #creates a list with the duplicates
    return duplicate_list

apple_id = frequency_var(Apple_data,0)
print(apple_id)

apple_track = frequency_var(Apple_data,1)
print(apple_track) #There are two duplicates - to Investigate


[]
[]


In [22]:
print(ios_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


# Delete Duplicates

__We have identified two duplicates for track_name 'Mannequin Challenge' and 'VR Roller Coaster'. Now, let's identify how many data entries there are and at what index in the dataset.__

In [24]:
index = 1

print(ios_header,'\n')

for i in Apple_data[1:]:
    if i[1] == 'Mannequin Challenge':
        print('Row at index:', index,'\n', i)
    elif i[1] == 'VR Roller Coaster':
        print('Row at index:', index, '\n', i)
       
    index += 1     

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'] 

Row at index: 2949 
 ['1173990889', 'Mannequin Challenge', '109705216', 'USD', '0.0', '668', '87', '3.0', '3.0', '1.4', '9+', 'Games', '37', '4', '1', '1']
Row at index: 4443 
 ['952877179', 'VR Roller Coaster', '169523200', 'USD', '0.0', '107', '102', '3.5', '3.5', '2.0.0', '4+', 'Games', '37', '5', '1', '1']
Row at index: 4464 
 ['1178454060', 'Mannequin Challenge', '59572224', 'USD', '0.0', '105', '58', '4.0', '4.5', '1.0.1', '4+', 'Games', '38', '5', '1', '1']
Row at index: 4832 
 ['1089824278', 'VR Roller Coaster', '240964608', 'USD', '0.0', '67', '44', '3.5', '4.0', '0.81', '4+', 'Games', '38', '0', '1', '1']


*Having looked at the data, it seems that the data have been captured at two different times, as per below:
     __'Mannequin Challenge'__ once at __rating_count_tot__ (no of ratings) of __107__ and once at __668__
     __'VR Roller Coaster'__ once at __rating_count_tot__ (no of ratings) of __67__ and once at __105__*
     
We will keep the data entries with the highest number of ratings, as it is a more recent and thus, accurate data entry. 

Rows __4832__ and __4464__ will be deleted. 

In [30]:
#del(Apple_data[4832])  #has been actioned
#del(Apple_data[4464])  #has been actioned

print(Apple_data[4832]) #successful deletion 
print(Apple_data[4464]) #successful deletion

['1062002361', 'LumaFX - infinite video effects', '13921280', 'USD', '2.99', '67', '11', '4.0', '4.5', '2.0.3', '4+', 'Photo & Video', '37', '5', '8', '1']
['1041406978', 'DOFUS Touch', '3366912', 'USD', '0.0', '104', '3', '4.0', '4.0', '1.9.28', '12+', 'Games', '37', '5', '6', '1']


# Google Store
## Identify Duplicates


I wrote the function __'frequency_var'__ earlier, that will be used on this occassion as well.

As a reminder, see the function below:

In [54]:
print('The are: ', len(Google_data)-1,' Google apps','\n') # -1 because we don't count the header in

duplicate_apps = []
unique_apps = []

for app in Google_data[1:]:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
    
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])


The are:  10840  Google apps 

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


In certain cases, we want to keep only one data entry per app. We could delete the extra entries at random but there must be a better way for sure. 

Let's see what differs within the data sets with the same name. 

In [60]:
print(android_header)
print('\n')

for x in Google_data[1:]:
    name = x[0]
    if name == 'Box':
        print(x)
    elif name == 'Zenefits':
        print(x)
             

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Zenefits', 'BUSINESS', '4.2', '296', '14M', '50,000+', 'Free', '0', 'Everyone', 'Business', 'June 15, 2018', '3.2.1', '4.1 and up']
['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Zenefits', 'BUSINESS', '4.2', '296', '14M', '50,000+', 'Free', '0', 'Everyone', 'Business', 'June 15, 2018', '3.2.1', '4.1 and up']


The above seem identical, let's look draw some more data points. It may be the case that the date differs or there are more installs, which indicates that we would be more inclined to keep some data points than others. 

In [61]:
print(android_header)
print('\n')

for x in Google_data[1:]:
    name = x[0]
    if name == 'Slack':
        print(x)
    elif name == 'Google Ads':
        print(x)
    elif name == 'Instagram':
        print(x)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Google Ads', 'BUSINESS', '4.3', '29313', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 30, 2018', '1.12.0', '4.0.3 and up']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Google Ads', 'BUSINESS', '4.3', '29313', '20M', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 30, 2018', '1.12.0', '4.0.3 and up']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '6657731

It seems that the data entries for the same app can differ in the number of reviews (e.g. Instagram, Slack) or can have the same number of reviews (e.g in our previous example for Box and Zenefits)

As such, we should aim at maintaining only one data point per app, with the highest number of ratings in case that number differs. That's because the higher the number of reviews, the more reliable the rating and our data will be. 

The way to do that is to create a dictionary, where each key contains the highest number of reviews for that particular app. And use that dictionary to create a completely new data set that will containg only one data entry per app. 

## Removing Duplicates 

So there are 10840  Google apps in the Google_data dataset, with 1181 duplicate apps
